[GISel] Explicitly disable BF16 tablegen patterns.

davemgreen opened the pull request:
We currently have an issue where bf16 patters can be used to match fp16 types, as GISel does not know about the difference between the two. This patch explicitly disables them to make sure that they are never used.

The opposite can also happen too, where fp16 patterns are used for operators that should be bf16. So this also changes any operations with bf16 types to now cause a fallback to SDAG.

The pass setup for GISel has been slightly adjusted to make sure that a verify pass does not get added between AMD-SDAG and SIFixSGPRCopiesPass, which otherwise can cause verifier issues when falling back.

sjoerdmeijer reviewed (approved):
This fixes a codegen/runtime error that we are seeing, as reported in https://github.com/llvm/llvm-project/pull/120363#issuecomment-2609616562

I am not a GISel guy, but I have applied and verified that this patch solves the problems that we are seeing, and the change looks very reasonable, so am confident to LGTM this. 

Thanks for the quick fix!

arsenm commented:
Do we have a tracking issue for bf16? Should mention to revert this there

madhur13490 commented:
I am relatively new to GISel intersection FP16. In the longer term, will we have a clean solution when GISel represents types natively? As Matt said above, an issue would be good to have and link to https://github.com/llvm/llvm-project/issues/115133?

davemgreen commented:
> Do we have a tracking issue for bf16? Should mention to revert this there

I couldn't see one for BF16 specifically, but I'll add a note to #115133 and FYI @tgymnich if you run into this. Hopefully it is fairly obvious that none of the patterns are importing or types work when we can use bf16 correctly. Let me know if you know of a better place, quite a lot has changed now..

Sorry - This is running into other issues, where the opposite happens for certain intrinsics and instructions with fp16 types are using bf16 patterns. As in https://godbolt.org/z/1KcvszrY7.  I have applied the fairly heavy hammer of forcing all IR with bf16 types to fall back to SDAG. This shouldn't be a problem for data-processing only instructions (load, stores, call params, shuffles, inserts, extract, select, etc there are a lot of them), but they are currently included to make sure everything is covered properly.

davemgreen committed:
[GISel] Explicitly disable BF16 tablegen patterns and codegen.

We currently have an issue where bf16 patters can be used to match fp16 types,
as GISel does not know about the difference between the two types. This patch
explicitly disables them to make sure that they are never used.

The opposite can also happen too, where fp16 patterns are used for operators
that should be bf16. So any operations with bf16 types now cause a fallback to
SDAG. For the moment this includes data-processing only instructions (loads,
stores, shufles, etc).

The pass setup for GISel has been slightly adjusted to make sure that a verify
pass does not get added between AMD-SDAG and SIFixSGPRCopiesPass, which
otherwise can cause verifier issues when falling back.

davemgreen committed:
Cleanup to avoid typesize-only operations.

davemgreen commented:
I've tried to clean this up to avoid typesize-only operations that are already tested and include just the instructions that cause problems. Hopefully it has not missed any, and calls are still excluded entirely.

mikaelholmen commented:
Hi @davemgreen

With this patch
```
llvm-lit -av test/CodeGen/AMDGPU/llvm.amdgcn.mfma.gfx950.ll
```
fails when llc is compiled with EXPENSIVE_CHECKS on:
```
# After Virtual Register Rewriter
********** INTERVALS **********
VGPR9_LO16 EMPTY
VGPR9_HI16 EMPTY
RegMasks:
********** MACHINEINSTRS **********
# Machine code for function test_mfma_f32_16x16x32_bf16_no_agpr__vgprcd: NoPHIs, TracksLiveness, NoVRegs, TiedOpsRewritten, FailedRegAlloc, TracksDebugUserValues
Function Live Ins: $sgpr4_sgpr5

0B	bb.0 (%ir-block.0):
	  liveins: $sgpr4_sgpr5
64B	  early-clobber renamable $sgpr8_sgpr9_sgpr10_sgpr11_sgpr12_sgpr13_sgpr14_sgpr15 = S_LOAD_DWORDX8_IMM_ec renamable $sgpr4_sgpr5, 52, 0 :: (dereferenceable invariant load (s256) from %ir.arg0.kernarg.offset, align 4, addrspace 4)
80B	  early-clobber renamable $sgpr0_sgpr1_sgpr2_sgpr3 = S_LOAD_DWORDX4_IMM_ec renamable $sgpr4_sgpr5, 84, 0 :: (dereferenceable invariant load (s128) from %ir.arg2.kernarg.offset, align 4, addrspace 4)
96B	  renamable $vgpr0_vgpr1_vgpr2_vgpr3 = COPY renamable $sgpr8_sgpr9_sgpr10_sgpr11
112B	  renamable $vgpr8 = V_MOV_B32_e32 0, implicit $exec
128B	  early-clobber renamable $sgpr6_sgpr7 = S_LOAD_DWORDX2_IMM_ec killed renamable $sgpr4_sgpr5, 36, 0 :: (dereferenceable invariant load (s64) from %ir.out.kernarg.offset, align 4, addrspace 4)
144B	  renamable $vgpr4_vgpr5_vgpr6_vgpr7 = COPY killed renamable $sgpr12_sgpr13_sgpr14_sgpr15
160B	  renamable $vgpr10_vgpr11_vgpr12_vgpr13 = COPY killed renamable $sgpr0_sgpr1_sgpr2_sgpr3
168B	  renamable $agpr0_agpr1_agpr2_agpr3 = COPY killed renamable $vgpr10_vgpr11_vgpr12_vgpr13
176B	  renamable $agpr0_agpr1_agpr2_agpr3 = V_MFMA_F32_16X16X32_BF16_e64 killed $vgpr0_vgpr1_vgpr2_vgpr3, killed $vgpr4_vgpr5_vgpr6_vgpr7, killed $agpr0_agpr1_agpr2_agpr3, 0, 0, 0, implicit $mode, implicit $exec
184B	  renamable $vgpr0_vgpr1_vgpr2_vgpr3 = COPY killed renamable $agpr0_agpr1_agpr2_agpr3
192B	  GLOBAL_STORE_DWORDX4_SADDR killed renamable $vgpr8, killed renamable $vgpr0_vgpr1_vgpr2_vgpr3, killed renamable $sgpr6_sgpr7, 0, 0, implicit $exec :: (store (s128) into %ir.out.load, addrspace 1)
208B	  S_ENDPGM 0

# End machine code for function test_mfma_f32_16x16x32_bf16_no_agpr__vgprcd.

*** Bad machine code: isRenamable set on reserved register ***
- function:    test_mfma_f32_16x16x32_bf16_no_agpr__vgprcd
- basic block: %bb.0  (0x55dab89e8b30) [0B;224B)
- instruction: 168B	renamable $agpr0_agpr1_agpr2_agpr3 = COPY killed renamable $vgpr10_vgpr11_vgpr12_vgpr13
- operand 0:   renamable $agpr0_agpr1_agpr2_agpr3

*** Bad machine code: isRenamable set on reserved register ***
- function:    test_mfma_f32_16x16x32_bf16_no_agpr__vgprcd
- basic block: %bb.0  (0x55dab89e8b30) [0B;224B)
- instruction: 176B	renamable $agpr0_agpr1_agpr2_agpr3 = V_MFMA_F32_16X16X32_BF16_e64 killed $vgpr0_vgpr1_vgpr2_vgpr3, killed $vgpr4_vgpr5_vgpr6_vgpr7, killed $agpr0_agpr1_agpr2_agpr3, 0, 0, 0, implicit $mode, implicit $exec
- operand 0:   renamable $agpr0_agpr1_agpr2_agpr3

*** Bad machine code: isRenamable set on reserved register ***
- function:    test_mfma_f32_16x16x32_bf16_no_agpr__vgprcd
- basic block: %bb.0  (0x55dab89e8b30) [0B;224B)
- instruction: 184B	renamable $vgpr0_vgpr1_vgpr2_vgpr3 = COPY killed renamable $agpr0_agpr1_agpr2_agpr3
- operand 1:   killed renamable $agpr0_agpr1_agpr2_agpr3
LLVM ERROR: Found 3 machine code errors.
```

Can also be seen if you use a normal build and just add "-verify-machineinstrs" to the command on line 3:
```llc -mtriple=amdgcn -mcpu=gfx950 -global-isel=1 -global-isel-abort=2 < test/CodeGen/AMDGPU/llvm.amdgcn.mfma.gfx950.ll -verify-machineinstrs```

```

davemgreen commented:
Thanks for the heads up - it looks like after we fall back from global-isel to SDAG, the verifier gets called, which calls getReservedRegs which uses SIMachineFunctionInfo::usesAGPRs which caches the result of UsesAGPRs. Because we have just fallen-back the function is empty and it incorrectly gets cached to false. This then causes a regallocation failure. One way around it might be to make sure we do not call getReservedRegs on an empty function in the verifier, but it may be better to reset that cached value somewhere, I'm just not sure where.

davemgreen commented:
New, hopefully better idea in #124799
